File: scip-java/src/main/scala/com/sourcegraph/scip_java/buildtools/GradleBuildTool.scala
package com.sourcegraph.scip_java.buildtools

import java.nio.charset.StandardCharsets
import java.nio.file._

import scala.collection.mutable.ListBuffer
import scala.util.Properties
import scala.util.Try

import com.sourcegraph.io.DeleteVisitor
import com.sourcegraph.scip_java.Embedded
import com.sourcegraph.scip_java.commands.IndexCommand
import os.CommandResult

class GradleBuildTool(index: IndexCommand) extends BuildTool("Gradle", index) {

  override def usedInCurrentDirectory(): Boolean = {
    val gradleFiles = List(
      "settings.gradle",
      "gradlew",
      "build.gradle",
      "build.gradle.kts"
    )
    gradleFiles.exists(name =>
      Files.isRegularFile(index.workingDirectory.resolve(name))
    )
  }

  override def generateScip(): Int = {
    val gradleResult = generateSemanticdb()
    if (gradleResult.exitCode == 0) {
      reportMissingSemanticdbOutput()
    }
    BuildTool.generateScipFromTargetroot(gradleResult, targetroot, index)
  }

  /**
   * Diagnose the case where Gradle finished successfully but our SemanticDB
   * compiler plugin never produced any `.semanticdb` files. This used to be
   * silently rescued by a `-javaagent` fallback; now it surfaces as a clear
   * error pointing at the two known causes.
   */
  private def reportMissingSemanticdbOutput(): Unit = {
    if (containsFileWithSuffix(targetroot, ".semanticdb"))
      return
    if (!containsFileWithSuffix(index.workingDirectory, ".class"))
      // Project produced no compiled JVM output — nothing to index, stay quiet.
      return
    index
      .app
      .reporter
      .error(
        s"""scip-java: Gradle finished successfully but produced no SemanticDB output in $targetroot.
           |
           |This means our SemanticDB compiler plugin was not attached to one or more JavaCompile tasks. Two known causes:
           |
           |  1. The 'compileOnly' configuration was already resolved before our init script ran.
           |     Check the Gradle output above for warnings of the form:
           |       "scip-java: failed to attach SemanticDB compiler plugin to project '<name>'"
           |     Workaround: apply the SemanticDB plugin earlier (e.g. via a settings plugin),
           |     or restructure the build so that 'compileOnly' is not resolved at evaluation time.
           |
           |  2. Another Gradle plugin is replacing the compiler arguments we add (rather than appending).
           |     Verify with:  ./gradlew compileJava --info | grep -- '-Xplugin:semanticdb'
           |     If '-Xplugin:semanticdb' is missing from the printed javac command, another plugin
           |     is overwriting JavaCompile.options.compilerArgs.
           |""".stripMargin
      )
  }

  private def containsFileWithSuffix(root: Path, suffix: String): Boolean =
    Files.isDirectory(root) &&
      Try {
        val stream = Files.find(
          root,
          Integer.MAX_VALUE,
          (p, attrs) =>
            attrs.isRegularFile && p.getFileName.toString.endsWith(suffix)
        )
        try stream.findFirst().isPresent
        finally stream.close()
      }.getOrElse(false)

  def targetroot: Path = index.finalTargetroot(defaultTargetroot)

  private def defaultTargetroot: Path = Paths.get(
    "build",
    "semanticdb-targetroot"
  )
  private def generateSemanticdb(): CommandResult = {
    val gradleWrapper: Path = index
      .workingDirectory
      .resolve(
        if (Properties.isWin)
          "gradlew.bat"
        else
          "gradlew"
      )
    val gradleCommand: String =
      if (
        Files.isRegularFile(gradleWrapper) && Files.isExecutable(gradleWrapper)
      )
        gradleWrapper.toString
      else
        "gradle"

    TemporaryFiles.withDirectory(index) { tmp =>
      runCompileCommand(tmp, gradleCommand)
    }
  }

  private def runCompileCommand(
      tmp: Path,
      gradleCommand: String
  ): CommandResult = {
    val script = initScript(tmp).toString
    val buildCommand = ListBuffer.empty[String]
    buildCommand += gradleCommand
    buildCommand += "--no-daemon"
    buildCommand += "--init-script"
    buildCommand += script
    buildCommand += "-Pkotlin.compiler.execution.strategy=in-process"
    buildCommand += s"-Dsemanticdb.targetroot=$targetroot"
    buildCommand ++=
      index.finalBuildCommand(
        List("clean", "scipPrintDependencies", "scipCompileAll")
      )

    Files.walkFileTree(targetroot, new DeleteVisitor())
    val result = index.process(buildCommand, env = Map("TERM" -> "dumb"))
    Embedded
      .reportUnexpectedJavacErrors(index.app.reporter, tmp)
      .getOrElse(result)
  }

  private def scipJavaDependencies = "scipJavaDependencies"

  private def initScript(tmp: Path): Path = {
    val pluginpath = Embedded.semanticdbJar(tmp)
    val gradlePluginPath = Embedded.gradlePluginJar(tmp)
    val semanticdbKotlincPath = Embedded.semanticdbKotlincJar(tmp)
    val dependenciesPath = targetroot.resolve("dependencies.txt")
    Files.deleteIfExists(dependenciesPath)

    val script =
      s"""
         | initscript {
         |     dependencies{ 
         |         classpath(files("${gradlePluginPath}"))
         |     }
         | }
         |
         | import com.sourcegraph.gradle.semanticdb.SemanticdbGradlePlugin
         |
         | allprojects {
         |   project.ext["semanticdbTarget"] = "$targetroot"
         |   project.ext["javacPluginJar"] = "$pluginpath"
         |   project.ext["dependenciesOut"] = "$dependenciesPath"
         |   project.ext["semanticdbKotlincJar"] = "$semanticdbKotlincPath"
         |   apply plugin: SemanticdbGradlePlugin
         | }
      """.stripMargin.trim

    Files.write(
      tmp.resolve("init-script.gradle"),
      script.getBytes(StandardCharsets.UTF_8)
    )
  }

}

File: scip-java/src/main/kotlin/com/sourcegraph/scip_java/buildtools/GradleBuildTool.kt
package com.sourcegraph.scip_java.buildtools

import com.sourcegraph.io.DeleteVisitor
import com.sourcegraph.scip_java.Embedded
import com.sourcegraph.scip_java.commands.IndexCommand
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

class GradleBuildTool(index: IndexCommand) : BuildTool("Gradle", index) {

    override fun usedInCurrentDirectory(): Boolean {
        val gradleFiles = listOf("settings.gradle", "gradlew", "build.gradle", "build.gradle.kts")
        return gradleFiles.any { name -> Files.isRegularFile(index.workingDirectory.resolve(name)) }
    }

    override fun generateScip(): Int {
        val gradleResult = runBuild()
        if (gradleResult.exitCode == 0) {
            reportMissingScipOutput()
        }
        return generateScipFromTargetroot(gradleResult, targetroot(), index)
    }

    /**
     * Diagnose the case where Gradle finished successfully but our SCIP
     * compiler plugin never produced any `.scip` shards. This used to be
     * silently rescued by a `-javaagent` fallback; now it surfaces as a
     * clear error pointing at the two known causes.
     */
    private fun reportMissingScipOutput() {
        if (containsFileWithSuffix(targetroot(), ".scip")) return
        if (!containsFileWithSuffix(index.workingDirectory, ".class")) {
            // Project produced no compiled JVM output — nothing to index, stay quiet.
            return
        }
        index.app.reporter.error(
            """scip-java: Gradle finished successfully but produced no SCIP shards in ${targetroot()}.

This means our SCIP compiler plugin was not attached to one or more JavaCompile tasks. Two known causes:

  1. The 'compileOnly' configuration was already resolved before our init script ran.
     Check the Gradle output above for warnings of the form:
       "scip-java: failed to attach SCIP compiler plugin to project '<name>'"
     Workaround: apply the SCIP plugin earlier (e.g. via a settings plugin),
     or restructure the build so that 'compileOnly' is not resolved at evaluation time.

  2. Another Gradle plugin is replacing the compiler arguments we add (rather than appending).
     Verify with:  ./gradlew compileJava --info | grep -- '-Xplugin:scip'
     If '-Xplugin:scip' is missing from the printed javac command, another plugin
     is overwriting JavaCompile.options.compilerArgs.
""",
        )
    }

    private fun containsFileWithSuffix(root: Path, suffix: String): Boolean {
        if (!Files.isDirectory(root)) return false
        return try {
            Files.find(
                root,
                Integer.MAX_VALUE,
                { p, attrs -> attrs.isRegularFile && p.fileName.toString().endsWith(suffix) },
            ).use { stream -> stream.findFirst().isPresent }
        } catch (_: Exception) {
            false
        }
    }

    fun targetroot(): Path = index.finalTargetroot(defaultTargetroot)

    private val defaultTargetroot: Path = Paths.get("build", "scip-targetroot")

    private fun runBuild(): ProcessResult {
        val gradleWrapperName =
            if (System.getProperty("os.name").lowercase().contains("win")) "gradlew.bat" else "gradlew"
        val gradleWrapper = index.workingDirectory.resolve(gradleWrapperName)
        val gradleCommand =
            if (Files.isRegularFile(gradleWrapper) && Files.isExecutable(gradleWrapper))
                gradleWrapper.toString()
            else "gradle"
        return TemporaryFiles.withDirectory(index) { tmp ->
            runCompileCommand(tmp, gradleCommand)
        }
    }

    private fun runCompileCommand(tmp: Path, gradleCommand: String): ProcessResult {
        val script = initScript(tmp).toString()
        val cmd = mutableListOf<String>()
        cmd += gradleCommand
        cmd += "--no-daemon"
        cmd += "--init-script"
        cmd += script
        cmd += "-Pkotlin.compiler.execution.strategy=in-process"
        cmd += "-Dscip.targetroot=${targetroot()}"
        cmd += index.finalBuildCommand(listOf("clean", "scipPrintDependencies", "scipCompileAll"))

        if (Files.exists(targetroot())) {
            Files.walkFileTree(targetroot(), DeleteVisitor())
        }
        val result = index.app.runProcess(cmd, env = mapOf("TERM" to "dumb"))
        return Embedded.reportUnexpectedJavacErrors(index.app.reporter, tmp) ?: result
    }

    private fun initScript(tmp: Path): Path {
        val pluginpath = Embedded.scipJar(tmp)
        val gradlePluginPath = Embedded.gradlePluginJar(tmp)
        val scipKotlincPath = Embedded.scipKotlincJar(tmp)
        val dependenciesPath = targetroot().resolve("dependencies.txt")
        Files.deleteIfExists(dependenciesPath)

        val script =
            """
             initscript {
                 dependencies{ 
                     classpath(files("${gradlePluginPath}"))
                 }
             }

             import com.sourcegraph.gradle.scip.ScipGradlePlugin

             allprojects {
               project.ext["scipTarget"] = "${targetroot()}"
               project.ext["javacPluginJar"] = "$pluginpath"
               project.ext["dependenciesOut"] = "$dependenciesPath"
               project.ext["scipKotlincJar"] = "$scipKotlincPath"
               apply plugin: ScipGradlePlugin
             }
            """.trimIndent()

        val out = tmp.resolve("init-script.gradle")
        Files.write(out, script.toByteArray(StandardCharsets.UTF_8))
        return out
    }
}
